### PR TITLE
Allow timescale to be a non-primary rails db

### PR DESCRIPTION
**Pull request by segiddins**
Enables rubygems.org to use timescale as a secondary db only for gem downloads: https://github.com/rubygems/rfcs/pull/48

Consumed in https://github.com/rubygems/rubygems.org/pull/3560

### CODE REVIEW

File: lib/timescaledb/rails/extensions/active_record/schema_dumper.rb
Comment (segiddins) [inline lib/timescaledb/rails/extensions/active_record/schema_dumper.rb:23]: continuous aggregates can depend on each other, and we need to make sure they're topologically ordered in the dumped schema

File: lib/timescaledb/rails/extensions/active_record/schema_dumper.rb
Comment (segiddins) [inline lib/timescaledb/rails/extensions/active_record/schema_dumper.rb:159]: only write to the schema file for the correct DB

File: lib/timescaledb/rails/extensions/active_record/schema_statements.rb
Comment (segiddins) [inline lib/timescaledb/rails/extensions/active_record/schema_statements.rb:121]: allows for re-running the same migration idempotently

File: lib/timescaledb/rails/model/hyperfunctions.rb
Comment (segiddins) [inline lib/timescaledb/rails/model/hyperfunctions.rb:13]: allows keeping the same shape of data in the resulting query as the source

File: lib/timescaledb/rails/model/hyperfunctions.rb
Comment (segiddins) [inline lib/timescaledb/rails/model/hyperfunctions.rb:23]: needed for when the alias is also a cloumn in the source table

File: lib/timescaledb/rails/models/chunk.rb
Comment (segiddins) [inline lib/timescaledb/rails/models/chunk.rb:6]: need to allow the user to specify an abstract superclass, since that's the level that connections are configured

File: lib/timescaledb/rails/railtie.rb
Comment (segiddins) [inline lib/timescaledb/rails/railtie.rb:11]: need to use `to_prepare` so integrating apps have a change to customize the record base

File: lib/timescaledb/rails/models/chunk.rb
Comment (santiagodoldan) [reply lib/timescaledb/rails/models/chunk.rb:6]: Testing new changes I merged today into main, using `bin/console` generates this error

```
uninitialized constant Timescaledb::Rails::Railtie (NameError)
```

**Comment by santiagodoldan** [comment]
@segiddins Thank you so much for your contribution, it's awesome to hear that rubygems will use the gem 🚀

File: lib/timescaledb/rails/extensions/active_record/schema_statements.rb
Comment (segiddins) [reply lib/timescaledb/rails/extensions/active_record/schema_statements.rb:121]: i think it's similar to the `force:` parameter to `create_table`.
I added this and set it to true because we're creating multiple CAs in one migration, and if it fails halfway through, you can't re-run it, since you can't create a materialized view that already exists

File: lib/timescaledb/rails/models/chunk.rb
Comment (segiddins) [reply lib/timescaledb/rails/models/chunk.rb:6]: That makes sense, can do that! Will rebase main & see what's changed there

File: lib/timescaledb/rails/extensions/active_record/schema_dumper.rb
Comment (segiddins) [reply lib/timescaledb/rails/extensions/active_record/schema_dumper.rb:36]: ah gotcha, i thi kk you're right

File: lib/timescaledb/rails/extensions/active_record/schema_statements.rb
Comment (santiagodoldan) [reply lib/timescaledb/rails/extensions/active_record/schema_statements.rb:121]: Ahhh, I forgot the `disable_ddl_transaction!` for continuous aggregates, the re-run part makes sense now, thanks for explaining it 👍

File: lib/timescaledb/rails/extensions/active_record/postgresql_database_tasks.rb
Comment (santiagodoldan) [inline lib/timescaledb/rails/extensions/active_record/postgresql_database_tasks.rb:159]: It would be great to find a way to extract this code and avoid the duplication from here and https://github.com/crunchloop/timescaledb-rails/pull/58/files#diff-fcf7b67734152346dd66ba227b705d281ef8010edfe237b18d3d24d47a1801eaL152-R165

File: lib/timescaledb/rails/extensions/active_record/schema_statements.rb
Comment (segiddins) [reply lib/timescaledb/rails/extensions/active_record/schema_statements.rb:134]: I think we do need the name to match up with the kwarg name in `create_continuous_aggregate`, since we forward along all the args

Comment (santiagodoldan) [reply lib/timescaledb/rails/extensions/active_record/schema_statements.rb:134]: Good point, as it is a named argument it needs to match, makes sense

File: lib/timescaledb/rails/models/application_record.rb
Comment (santiagodoldan) [inline lib/timescaledb/rails/models/application_record.rb:20]: This looks much better now, thank you 💯 . I would remove the `Hypertable.table_exists?` check here, I'd only check the connection here.

File: lib/timescaledb/rails/railtie.rb
Comment (santiagodoldan) [reply lib/timescaledb/rails/railtie.rb:9]: Would be ok to revert this change now?

**Comment by santiagodoldan** [comment]
@segiddins One last thing, could you follow conventional commits specs -> https://github.com/crunchloop/timescaledb-rails/blob/main/CONTRIBUTING.md#pull-requests?

**Comment by segiddins** [comment]
@santiagodoldan feel free to change the commit message when squash-merging

**Comment by santiagodoldan** [comment]
@segiddins Sure, that works too, let me know about my last 2 comments and I will approve your PR when they are done, I'll push a new version to rubygems if everything works as expected 🚀

**Comment by santiagodoldan** [comment]
@segiddins Hey, let me know if my comments make sense to you, so I can resolve them and merge your changes 🚀

File: lib/timescaledb/rails/models/application_record.rb
Comment (segiddins) [reply lib/timescaledb/rails/models/application_record.rb:20]: So I think the issue with that is if someone has the gem installed and has not configured a custom connection for it -- we will assume that the primary conenction is for timescale, even if it is e.g. postgres

Comment (santiagodoldan) [reply lib/timescaledb/rails/models/application_record.rb:20]: Reading my comment again, I was not clear enough, sorry about that. My suggestion was to move it back to where it was before because it seems weird to have a reference of an ApplicationRecord child here in the parent, what do you think? We would only check the connection here (which based on the name makes sense) and then  in the timescale_enabled? helper method, we would check the connection + if the hypertable table exists, sounds ok to you?
```ruby
def timescale_enabled?
  ApplicationRecord.timescale_connection?(connection) && Timescaledb::Rails::Hypertable.table_exists?
end
```

Comment (segiddins) [reply lib/timescaledb/rails/models/application_record.rb:20]: Ah yup that makes sense. Can make that change tonight.

Comment (santiagodoldan) [reply lib/timescaledb/rails/models/application_record.rb:20]: I'll try to publish a new version as soon as we get this merged 💪

Comment (santiagodoldan) [reply lib/timescaledb/rails/models/application_record.rb:20]: @segiddins Let me know if there is anything I can help you with, I'd love to merge this and publish a new version 💯

**Comment by santiagodoldan** [comment]
@segiddins Looks like `ruby2_keywords` was introduced in ruby 2.7

**Review by santiagodoldan — APPROVED** [review]

**Comment by santiagodoldan** [comment]
@segiddins Thanks for your contribution 💯

**Pull request merged**
